File: engine/swig/python/primitives.c
/************************************************************************************/
/* Copyright (c) 2009 The Department of Arts and Culture,                           */
/* The Government of the Republic of South Africa.                                  */
/*                                                                                  */
/* Contributors:  Meraka Institute, CSIR, South Africa.                             */
/*                                                                                  */
/* Permission is hereby granted, free of charge, to any person obtaining a copy     */
/* of this software and associated documentation files (the "Software"), to deal    */
/* in the Software without restriction, including without limitation the rights     */
/* to use, copy, modify, merge, publish, distribute, sublicense, and/or sell        */
/* copies of the Software, and to permit persons to whom the Software is            */
/* furnished to do so, subject to the following conditions:                         */
/* The above copyright notice and this permission notice shall be included in       */
/* all copies or substantial portions of the Software.                              */
/*                                                                                  */
/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR       */
/* IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,         */
/* FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE      */
/* AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER           */
/* LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,    */
/* OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN        */
/* THE SOFTWARE.                                                                    */
/*                                                                                  */
/************************************************************************************/
/*                                                                                  */
/* AUTHOR  : Aby Louw                                                               */
/* DATE    : November 2009                                                          */
/*                                                                                  */
/************************************************************************************/
/*                                                                                  */
/* C convenience functions for primitives SInt, SFloat, SString.                    */
/*                                                                                  */
/*                                                                                  */
/************************************************************************************/


%define sobject_2_pyobject_DOCSTRING
"""
sobject_2_pyobject(object, own)

Convert a Speect ``SObject`` type object to a Python object. This can be
use full for conversions where either type is required an is used extensively
internally and in the plug-in implementations.

The ``own`` ownership flag specifies if Python must take ownership of the object,
and hence can delete it. If not then Speect controls the memory of the object.

:param object: The Speect object to convert.
:type key: SObject
:param own: Ownership flag.
:type own: bool
:return: A Python representation of the given object
:rtype: Python type object.
"""
%enddef

%feature("autodoc", sobject_2_pyobject_DOCSTRING) sobject_2_pyobject;


%define pyobject_2_sobject_DOCSTRING
"""
pyobject_2_sobject(pobject)

Convert a Python type object to a Speect ``SObject`` object. This can be
use full for conversions where either type is required an is used extensively
internally and in the plug-in implementations.

Note that for basic types (int, float, string) Speect creates a new type. For
other types, Speect creates an ``SObject``. This function will only work with
basic types if no external plug-ins are loaded.

:param pobject: The Python object to convert.
:return: A Speect representation of the given object
:rtype: ``SObject`` type object.
"""
%enddef

%feature("autodoc", pyobject_2_sobject_DOCSTRING) pyobject_2_sobject;



/*
 * Do not delete these delimiters, required for SWIG
 */
%inline
%{
	PyObject *sobject_2_pyobject(const SObject *object, s_erc *error, s_bool own)
	{
		PyObject *pobject;
		const char *type;
		int s_comp;
		swig_type_info *info;


		if (object == NULL)
			Py_RETURN_NONE;

		type = SObjectType(object, error);
		if (*error != S_SUCCESS)
			return NULL;

		s_comp = s_strcmp(type, "SInt", error);
		if (*error != S_SUCCESS)
			return NULL;

		if (s_comp == 0)
		{
			sint32 val;


			val = SObjectGetInt(object, error);
			if (*error != S_SUCCESS)
				return NULL;

			pobject = PyInt_FromLong((long)val);
			if (own == TRUE)
				S_DELETE(object, "sobject_2_pyobject", error);

			return pobject;
		}

		s_comp = s_strcmp(type, "SFloat", error);
		if (*error != S_SUCCESS)
			return NULL;

		if (s_comp == 0)
		{
			float val;


			val = SObjectGetFloat(object, error);
			if (*error != S_SUCCESS)
				return NULL;

			pobject = PyFloat_FromDouble((double)val);
			if (own == TRUE)
				S_DELETE(object, "sobject_2_pyobject", error);

			return pobject;
		}

		s_comp = s_strcmp(type, "SString", error);
		if (*error != S_SUCCESS)
			return NULL;

		if (s_comp == 0)
		{
			const char *val;
			size_t slen;


			val = SObjectGetString(object, error);
			if (*error != S_SUCCESS)
				return NULL;

			slen = s_strsize(val, error);
			if (*error != S_SUCCESS)
				return NULL;

			pobject = PyString_FromStringAndSize(val, slen);
			if (own == TRUE)
				S_DELETE(object, "sobject_2_pyobject", error);

			return pobject;
		}

		s_comp = s_strcmp(type, "SPyObject", error);
		if (*error != S_SUCCESS)
			return NULL;

		if (s_comp == 0)
		{
			/* note that the get method implementation of the
			 * SPyObject increased the reference count of the returned
			 * python object */
			pobject = (PyObject*)SObjectGetVoid(object, error);
			if (*error != S_SUCCESS)
				return NULL;

			if (own == TRUE)
				S_DELETE(object, "sobject_2_pyobject", error);

			return pobject;
		}

		/* cycle through object inheritance hierarchy,
		 * maybe we find something
		 */
		{
			char               *cname;
			char               *c;
			char               *cp;
			const char         *inherit_hier;
			size_t             len;
			char               *full_type_name;

			inherit_hier = SObjectInheritance(object, error);
			if (*error != S_SUCCESS)
				return NULL;

			cname = s_strdup(inherit_hier, error);
			if (*error != S_SUCCESS)
				return NULL;

			cp = cname;
			do
			{
				c = s_strtok_r(NULL, ":", &cp, error);

				if (S_CHK_ERR(error, S_CONTERR,
							  "sobject_2_pyobject",
							  "Failed to extract token from class name"))
				{
					S_FREE(cname);
					return NULL;
				}

				if (c == NULL)
				{
					S_FREE(cname);
					break;
				}

				len = s_strzsize(c, error) + 2;
				if (*error != S_SUCCESS)
				{
					S_FREE(cname);
					return NULL;
				}

				full_type_name = S_CALLOC(char, len);
				if (full_type_name == NULL)
				{
					S_FREE(cname);
					S_FTL_ERR(error, S_MEMERROR,
							  "sobject_2_pyobject",
							  "Failed to allocate memory for 'char' object");
					return NULL;
				}

				s_strcat(full_type_name, c, error);
				if (*error != S_SUCCESS)
				{
					S_FREE(full_type_name);
					S_FREE(cname);
					return NULL;
				}

				s_strcat(full_type_name, " *", error);
				if (*error != S_SUCCESS)
				{
					S_FREE(full_type_name);
					S_FREE(cname);
					return NULL;
				}

				info = SWIG_TypeQuery(full_type_name);
				S_FREE(full_type_name);

				if (info != NULL)
				{
					S_FREE(cname);
					break;
				}

			} while (c != NULL);
		}

		if (info != NULL)
		{
			if (own == TRUE)
			{
				pobject = SWIG_NewPointerObj(SWIG_as_voidptr(object),
											 info, 1);
			}
			else
			{
				pobject = SWIG_NewPointerObj(SWIG_as_voidptr(object),
											 info, 0 |0);
			}
		}
		else
		{
			/* don't know this object type, use SObject */
			if (own == TRUE)
			{
				pobject = SWIG_NewPointerObj(SWIG_as_voidptr(object),
											 SWIGTYPE_p_SObject, 1);
			}
			else
			{
				pobject = SWIG_NewPointerObj(SWIG_as_voidptr(object),
											 SWIGTYPE_p_SObject, 0 |0);
			}
		}

		if (own == TRUE)
			S_DELETE(object, "sobject_2_pyobject", error);

		return pobject;
	}


	SObject *pyobject_2_sobject(PyObject *pobject, s_erc *error)
	{
		SObject *object;
		int res;
		void *argp;


		if (pobject == Py_None)
			return NULL;

		res = SWIG_ConvertPtr(pobject, &argp,SWIGTYPE_p_SObject, SWIG_POINTER_DISOWN|0);
		if (SWIG_IsOK(res))
		{
			/* it's a speect object */
			object = (SObject*)argp;
			return object;
		}

		/* it's a python object */
		if (PyObject_IsInstance(pobject, (PyObject*)&PyInt_Type))
		{
			sint32 tmp;


			tmp = (sint32)PyLong_AsLong(pobject);
			object = SObjectSetInt(tmp, error);
			if (*error != S_SUCCESS)
				return NULL;

			return object;
		}

		if (PyObject_IsInstance(pobject, (PyObject*)&PyFloat_Type))
		{
			float tmp;


			tmp = (float)PyFloat_AsDouble(pobject);
			object = SObjectSetFloat(tmp, error);
			if (*error != S_SUCCESS)
				return NULL;

			return object;
		}

		if (PyObject_IsInstance(pobject, (PyObject*)&PyString_Type))
		{
			const char *tmp;


			tmp = PyString_AsString(pobject);
			object = SObjectSetString(tmp, error);
			if (*error != S_SUCCESS)
				return NULL;

			return object;
		}

		if (PyObject_IsInstance(pobject, (PyObject*)&PyUnicode_Type))
		{
			PyObject *ustring;
			const char *tmp;


			ustring = PyUnicode_AsUTF8String(pobject);
			if (ustring == NULL)
			{
				S_CTX_ERR(error, S_FAILURE,
						  "pyobject_2_sobject",
						  "Call to \"PyUnicode_AsUTF8String\" failed");
				return NULL;
			}

			tmp = PyString_AsString(ustring);
			object = SObjectSetString(tmp, error);
			if (*error != S_SUCCESS)
				return NULL;

			Py_XDECREF(ustring);
			return object;
		}

		if (PyObject_IsInstance(pobject, (PyObject*)&PyList_Type))
		{
			SList *list;
			int count;
			int i;


			count = PyList_Size(pobject);
			if (count == 0)
				return NULL;

			list = (SList*)S_NEW("SListList", error);
			if (*error != S_SUCCESS)
				goto slist_fail;

			SListListInit(&list, error);
			if (*error != S_SUCCESS)
				goto slist_fail;

			for (i = 0; i < count; i++)
			{
				PyObject *p = PyList_GetItem(pobject, i);


				if (PyString_Check(p))
				{
					const char *tmp;
					SObject *obj;


					tmp = PyString_AsString(p);
					obj = SObjectSetString(tmp, error);
					if (*error != S_SUCCESS)
						goto slist_fail;

					SListAppend(list, obj, error);
					if (*error != S_SUCCESS)
						goto slist_fail;
				}
				else if (PyUnicode_Check(p))
				{
					const char *tmp;
					SObject *obj;
					PyObject *ustring;


					ustring = PyUnicode_AsUTF8String(p);
					if (ustring == NULL)
					{
						S_CTX_ERR(error, S_FAILURE,
								  "pyobject_2_sobject",
								  "Call to \"PyUnicode_AsUTF8String\" failed");
						goto slist_fail;
					}

					tmp = PyString_AsString(p);
					obj = SObjectSetString(tmp, error);
					if (error != S_SUCCESS)
						goto slist_fail;

					SListAppend(list, obj, error);
					if (*error != S_SUCCESS)
						goto slist_fail;
				}
				else
				{
					PyErr_SetString(PyExc_TypeError,
									"object in list is not str/unicode type");
					goto slist_fail;
				}
			}

			goto slist_good;

		slist_fail:
			if (list != NULL)
				S_DELETE(list, "pyobject_2_sobject", error);
		slist_good:
			object = S_OBJECT(list);

			return object;
		}


		/* Differs from speect.sobject_2_pyobject
		 * Adds support for Python object
		 *
		 * Not a simple object, make a SPyObject SVoid type
		 * Needs speect.pyobject module to work.
		 *
		 */
		object = SObjectSetVoid("SPyObject", (void*)pobject, error);
		if (*error != S_SUCCESS)
			return NULL;

		return object;
	}


/*
 * Do not delete this delimiter, required for SWIG
 */
%}

File: engine/swig/python/primitives.py
######################################################################################
## Copyright (c) 2009 The Department of Arts and Culture,                           ##
## The Government of the Republic of South Africa.                                  ##
##                                                                                  ##
## Contributors:  Meraka Institute, CSIR, South Africa.                             ##
##                                                                                  ##
## Permission is hereby granted, free of charge, to any person obtaining a copy     ##
## of this software and associated documentation files (the "Software"), to deal    ##
## in the Software without restriction, including without limitation the rights     ##
## to use, copy, modify, merge, publish, distribute, sublicense, and#or sell        ##
## copies of the Software, and to permit persons to whom the Software is            ##
## furnished to do so, subject to the following conditions:                         ##
## The above copyright notice and this permission notice shall be included in       ##
## all copies or substantial portions of the Software.                              ##
##                                                                                  ##
## THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR       ##
## IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,         ##
## FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE      ##
## AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER           ##
## LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,    ##
## OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN        ##
## THE SOFTWARE.                                                                    ##
##                                                                                  ##
######################################################################################
##                                                                                  ##
## AUTHOR  : Aby Louw                                                               ##
## DATE    : November 2009                                                          ##
##                                                                                  ##
######################################################################################
##                                                                                  ##
## Python classes for primitives SInt, SFloat, SString and SVoid.                   ##
##                                                                                  ##
##                                                                                  ##
######################################################################################


%pythoncode
%{

class SInt(SObject):
    """
    A class for the I{int} primitive data type.
    """

    def __init__(self, val=None, object=None, owner=False):
        """
        Construct a new SInt object. The input can either be a
        Python int type, or a C Speect SInt type SObject.
        @param val: Integer value for new SInt object.
        @type val: int
        @param object: A pointer to a C-type Speect SInt type
        SObject (default = I{None}).
        @type object: I{PySwigObject SObject*}
        @param owner: I{True} if the Python SInt is the owner of the
        C-type Speect SObject (and can therefore delete it),
        otherwise I{False} (default = I{False}).
        @type owner: bool
        @return: A new SInt instance.
        @rtype: L{SInt}
        """

        if val is not None:
            # test if val is an int
            if isinstance(val, int):
                new_spct_object = py_sint_set(val)

                # this is a new object, so this instance is
                # the owner for now
                super(SInt, self).__init__(new_spct_object, True) 
            else:
                raise TypeError('Input argument \"val\" must be ' +
                                'of type \'int\'')
        elif object:
            if not py_sobject_is_type(object, "SInt"):
                raise TypeError('Input argument \"object\" must be ' +
                                'of type \'C Speect SInt\'')
            else:
                super(SInt, self).__init__(object, owner)
        else:
            raise RuntimeError('Invalid input arguments')


    def __repr__(self):
        """
        Return a string representation of the object.
        Calls SInt.get() internally.
        @return: A string representation of the object.
        @rtype: str
        """
        return str(self.get())


    def get(self):
        """
        Get the value of this SInt object.
        @return: The integer value of this SInt object instance.
        @rtype: int
        """

        spct_object = self._get_speect_object()
        if spct_object:
            return py_sint_get(spct_object)
        else:
            return None



class SFloat(SObject):
    """
    A class for the I{float} primitive data type.
    """

    def __init__(self, val=None, object=None, owner=False):
        """
        Construct a new SFloat object. The input can either be a
        Python int type, or a C Speect SFloat type SObject.
        @param val: Float value for new SFloat object.
        @type val: float
        @param object: A pointer to a C-type Speect SFloat type
        SObject (default = I{None}).
        @type object: I{PySwigObject SObject*}
        @param owner: I{True} if the Python SFloat is the owner of the
        C-type Speect SObject (and can therefore delete it),
        otherwise I{False} (default = I{False}).
        @type owner: bool
        @return: A new SFloat instance.
        @rtype: L{SFloat}
        """

        if val:
            # test if val is a float
            if isinstance(val, float) or isinstance(val, int):
                new_spct_object = py_sfloat_set(float(val))

                # this is a new object, so this instance is
                # the owner for now
                super(SFloat, self).__init__(new_spct_object, True) 
            else:
                raise TypeError('Input argument \"val\" must be ' +
                                'of type \'float\'')
        elif object:
            if not py_sobject_is_type(object, "SFloat"):
                raise TypeError('Input argument \"object\" must be ' +
                                'of type \'C Speect SFloat\'')
            else:
                super(SFloat, self).__init__(object, owner) 
        else:
            raise RuntimeError('Invalid input arguments')


    def __repr__(self):
        """
        Return a string representation of the object.
        Calls SFloat.get() internally.
        @return: A string representation of the object.
        @rtype: str
        """
        return str(self.get())


    def get(self):
        """
        Get the value of this SFloat object.
        @return: The float value of this SFloat object instance.
        @rtype: float
        """

        spct_object = self._get_speect_object()
        if spct_object:
            return py_sfloat_get(spct_object)
        else:
            return None        



class SString(SObject):
    """
    A class for the I{char*} primitive data type.
    """

    def __init__(self, val=None, object=None, owner=False):
        """
        Construct a new SString object. The input can either be a
        Python str/unicode type, or a C Speect SString type SObject.
        @param val: String value for new SString object.
        @type val: unicode or str
        @param object: A pointer to a C-type Speect SString type
        SObject (default = I{None}).
        @type object: I{PySwigObject SObject*}
        @param owner: I{True} if the Python SString is the owner of the
        C-type Speect SObject (and can therefore delete it),
        otherwise I{False} (default = I{False}).
        @type owner: bool
        @return: A new SString instance.
        @rtype: L{SString}
        """

        if val:
            # test if val is a string type
            if isinstance(val, str) or isinstance(val, unicode):
                new_spct_object = py_sstring_set(val.encode('utf-8'))

                # this is a new object, so this instance is
                # the owner for now
                super(SString, self).__init__(new_spct_object, True) 
            else:
                raise TypeError('Input argument \"val\" must be ' +
                                'of type \'unicode\' or \'str\'')
        elif object:
            if not py_sobject_is_type(object, "SString"):
                raise TypeError('Input argument \"object\" must be ' +
                                'of type \'C Speect SString\'')
            else:
                super(SString, self).__init__(object, owner) 
        else:
            raise RuntimeError('Invalid input arguments')


    def __repr__(self):
        """
        Return a string representation of the object.
        Calls SString.get() internally.
        @return: A string representation of the object.
        @rtype: str
        """
        return self.get()


    def get(self):
        """
        Get the value of this SString object.
        @return: The string value of this SString object instance.
        @rtype: unicode
        """

        spct_object = self._get_speect_object()
        if spct_object:
            string = py_sstring_get(spct_object)
            if string:
                return unicode(string, 'utf-8')
            else:
                return None
        else:
            return None



class SVoid(SObject):
    """
    A class for the I{void*} primitive data type.
    """

    def __init__(self, val=None, object=None, owner=False):
        """
        Construct a new SVoid object. The input can either be a
        Python object (anytype), or a C Speect SVoid type SObject.
        @param val: Any Python type object.
        @param object: A pointer to a C-type Speect SVoid type
        SObject (default = I{None}).
        @type object: I{PySwigObject SObject*}
        @param owner: I{True} if the Python SVoid is the owner of the
        C-type Speect SObject (and can therefore delete it),
        otherwise I{False} (default = I{False}).
        @type owner: bool
        @return: A new SVoid instance.
        @rtype: L{SVoid}
        """

        if val and not object:
            new_spct_object = py_svoid_set(val)
            # this is a new object, so this instance is
            # the owner for now
            super(SVoid, self).__init__(new_spct_object, True) 

        elif object and not val:
            if not py_sobject_is_type(object, "SVoid"):
                raise TypeError('Input argument \"object\" must be ' +
                                'of type \'C Speect SVoid\'')
            else:
                super(SVoid, self).__init__(object, owner) 
        else:
            raise RuntimeError('Invalid input arguments')


    def get(self):
        """
        Get the value of this SVoid object.
        @return: The Python object value of this SVoid object
        instance.
        @rtype: Python object
        """

        spct_object = self._get_speect_object()
        if spct_object:
            return py_svoid_get(spct_object)
        else:
            return None
%}
